### ISSUE TITLE
Support recognizing TypeScript files imported with `.js` suffix

### ISSUE THREAD

**Issue by mon-jai** [pzavolinsky/ts-unused-exports#214]
In TypeScript, it is valid to import TypeScript files with `.js` extension. TypeScript will match the coresponding `.ts` files if they exist. This behaviour is commonly used in Node.js projects as [`.js` file extensions are required in Node.js ESM imports](https://nodejs.org/api/esm.html#esm_mandatory_file_extensions).

It would be great if ts-unused-exports also recognize variable imported with `.js` suffix.

Related: https://github.com/evanw/esbuild/issues/1343#issuecomment-854055814, https://github.com/microsoft/TypeScript/issues/16577#issuecomment-703190339, https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937

## Example

In [this example project derived from `example/simple`](https://github.com/mon-jai/ts-unused-exports/tree/master/example/simple), `inc()` from `math.ts` was imported with `.js` extension.

https://github.com/pzavolinsky/ts-unused-exports/blob/ad208c946855f9f5456d3d20435a4282f946c72b/example/simple/app.ts#L1

And TypeScript recognize it sucessfully.

![56-18-22-235633](https://user-images.githubusercontent.com/91261297/154717100-42f6acec-fa92-46ab-a83c-987cd85848a5.png)

However, ts-unused-exports failed to recognize that the default import of`math.ts` was imported in `app.ts`.

![Screenshot 2022-02-19 000527](https://user-images.githubusercontent.com/91261297/154718960-670bd156-6dc6-484b-a643-aba2c3825007.png)

**Comment by mrseanryan** [comment]
hm I'm not sure we will support imports of `.js` files in this way.

That link seems to be discussing ECMAScript (the spec for JavaScript).

However `ts-unused-exports` currently processes TypeScript.

Besides that, I don't really understand why this feature would be needed - isn't it possible use all TypeScript with Node.js ?

Perhaps if you can explain the use case a bit more, that would help ...

**Comment by mon-jai** [comment]
Importing without `.js` extension won't work if the target of your `tsc` output is ES6 or above (e.g. developing an ESM npm package).

For example, in [this modified example/simple project](https://github.com/pzavolinsky/ts-unused-exports/commit/56939d1730404f57013d8e032ea9d5ba5b1b6f89), I am targeting ES6 for TypeScript output.

https://github.com/pzavolinsky/ts-unused-exports/blob/56939d1730404f57013d8e032ea9d5ba5b1b6f89/example/simple/tsconfig.json#L3-L4

Without having `.js` extension in import statements of your **TypeScript files**, the compiled output won't run, as [TypeScript won't add the extension for you](https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937).

```
$ tsc && node app.js
node:internal/process/esm_loader:94
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'D:\GitHub\ts-unused-exports\example\simple\math' imported from D:\GitHub\ts-unused-exports\example\simple\app.js
Did you mean to import ../math.js?
    at new NodeError (node:internal/errors:371:5)
    at finalizeResolution (node:internal/modules/esm/resolve:394:11)
    at moduleResolve (node:internal/modules/esm/resolve:944:10)
    at defaultResolve (node:internal/modules/esm/resolve:1041:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
    at link (node:internal/modules/esm/module_job:78:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v17.4.0
```

**Comment by mon-jai** [comment]
If TypeScript output target is ES6 or above (e.g. developing an ESM npm package) and **TypeScript source files** do not have `.js` extension in their import statements, the compiled output won't run on Node.js.

In fact, example/simple only works when the output format is set to commonjs. If I [modify it to output ESM instead](https://github.com/pzavolinsky/ts-unused-exports/commit/56939d1730404f57013d8e032ea9d5ba5b1b6f89),

https://github.com/pzavolinsky/ts-unused-exports/blob/56939d1730404f57013d8e032ea9d5ba5b1b6f89/example/simple/tsconfig.json#L3-L4

the compiled output won't run, as [TypeScript won't add the extension for you](https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937).

```
$ tsc && node app.js
node:internal/process/esm_loader:94
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'D:\GitHub\ts-unused-exports\example\simple\math' imported from D:\GitHub\ts-unused-exports\example\simple\app.js
Did you mean to import ../math.js?
    at new NodeError (node:internal/errors:371:5)
    at finalizeResolution (node:internal/modules/esm/resolve:394:11)
    at moduleResolve (node:internal/modules/esm/resolve:944:10)
    at defaultResolve (node:internal/modules/esm/resolve:1041:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
    at link (node:internal/modules/esm/module_job:78:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v17.4.0
```

**Comment by mon-jai** [comment]
I am developing an npm package in ESM format. As TypeScript won't add file extension for me [by design](https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937), which is [required in Node.js ESM imports](https://nodejs.org/api/esm.html#esm_mandatory_file_extensions), I have no other choice but adding `.js` extension to import statements, so the compiled output contains them and works on Node.js.

**Comment by ephys** [comment]
Specifying the `.js` extension is the only way to load TypeScript files when using Node with native ES Modules, as ESM requires the extension to be specified and as @mon-jai said, TypeScript decided to not modify import paths at all (not replacing `.ts` with `.js` and not adding `.js` if the extension is unspecified)

**Comment by bbeesley** [comment]
Currently having to disable this tool in all our repos since migrating to node16/esm. There are plenty of details [in the ts docs](https://www.typescriptlang.org/docs/handbook/esm-node.html) if you need to understand why we have to import like this now @mrseanryan .

**Comment by mrseanryan** [comment]
Hello - this is somewhat tricky to add support for...

Here is a PR that adds initial support:

https://github.com/pzavolinsky/ts-unused-exports/pull/255

If someone can try it out, that would be great ...

**Comment by mrseanryan** [comment]
related:

- do we need to support the new extensions:  mjs, cjs, mts, cts

ref:
- https://www.typescriptlang.org/docs/handbook/esm-node.html

**Comment by mrseanryan** [comment]
Released as 9.0.0 - thank you

**Comment by ThomasAribart** [comment]
@mrseanryan Hello and thanks again for this wonderful library 🙌

It seems the issue has not been resolved when `.js` files are used in conjunction with `tsconfig` aliases. Here's the configuration on [DynamoDB-Toolbox](https://github.com/dynamodb-toolbox/dynamodb-toolbox):

```
// tsconfig.json
{
  ...
  "compilerOptions": {
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "baseUrl": "src",
    "target": "es2019",
    "rootDir": "src",
    "outDir": "dist/types",
    "declaration": true,
    ...
    "paths": {
      "~/*": ["*"]
    }
  }
}
```

When I run `ts-unused-exports` I get a lot of false positives. I have tried:
- Removing a `.js` extension on a file with an absolute import (`from '~/someFolder/someFile'`): It worked as intended
- Using a `.js` extension on a file with a relative import (`from './someFile.js'`): It also worked as intended

But the conjunction of the two (`from '~/someFolder/someFile.js'`) did not.

Do you confirm that's the current behavior? Is it the desired one? Can I do something to help?

Many thanks, Thomas
